can/zynqmp-can/zynqmp-can.c

```diff
@@ -0,0 +1,552 @@
+/*
+ * Phoenix-RTOS
+ *
+ * ZynqMP CAN driver
+ *
+ * Copyright 2025 Phoenix Systems
+ * Author: Dariusz Sabala
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <sys/debug.h>
+#include <sys/file.h>
+#include <sys/interrupt.h>
+#include <sys/mman.h>
+#include <sys/msg.h>
+#include <sys/platform.h>
+#include <sys/stat.h>
+#include <sys/threads.h>
+
+#include <board_config.h>
+
+#include <posix/utils.h>
+
+#include <phoenix/arch/aarch64/zynqmp/zynqmp.h>
+
+#include "zynqmp-can-priv.h"
+#include "zynqmp-can-if.h"
+
+/* Macro to calculate the length of an array */
+#define ARRAY_LENGTH(array) (sizeof((array)) / sizeof((array)[0]))
+
+/**
+ * Watermark configuration for "TX FIFO empty elements above watermark" interrupt.
+ * Setting this to low values (e.g., 1) reduces the blocking time for the sending function
+ * but increases the frequency of interrupts.
+ */
+#define CAN_TX_FIFO_WATERMARK 4
+
+/* Default CAN baud rate in kbps */
+#define CAN_BAUDRATE_KBPS_DEFAULT (1000)
+
+/* Base addresses for ZynqMP CAN peripherals */
+#define ZYNQMP_CAN0_BASE (0x00FF060000)
+#define ZYNQMP_CAN1_BASE (0x00FF070000)
+
+/* CAN peripheral instances */
+enum { can_0 = 0,
+	can_1 = 1,
+	canInstancesCount = 2 };
+
+/* Structure representing the CAN peripheral registers */
+typedef struct {
+	uint32_t SRR;             /* Software Reset and Enable, Address offset:  0x0 */
```

Comment:
All fields should be volatile
```suggestion
	volatile uint32_t srr;             /* Software Reset and Enable, Address offset:  0x0 */
```